ixmp4/data/api/base.py

```diff
@@ -40,26 +60,63 @@ class BaseModel(PydanticBaseModel):
     model_config = ConfigDict(from_attributes=True)
 
 
-def df_to_dict(df: pd.DataFrame) -> dict:
+class DataFrameDict(TypedDict):
+    index: list[int] | list[str]
+    columns: list[str]
+    dtypes: list[str]
+    # NOTE This is deliberately slightly out of sync with DataFrame.data below
+    # (cf positioning of int and float) to demonstrate that only the one below seems to
+    # affect our tests by causing ValidationErrors
+    data: list[
+        list[
+            abstract.annotations.PrimitiveTypes
+            | datetime
+            | dict[str, Any]
+            # TODO should be able to remove this once PR#122 is merged
+            | list[float | int | str]
+            | None
+        ]
+    ]
+
+
+def df_to_dict(df: pd.DataFrame) -> DataFrameDict:
     columns = []
     dtypes = []
     for c in df.columns:
         columns.append(c)
         dtypes.append(df[c].dtype.name)
 
-    return {
-        "index": df.index.to_list(),
-        "columns": columns,
-        "dtypes": dtypes,
-        "data": df.values.tolist(),
-    }
+    return DataFrameDict(
+        index=df.index.to_list(),
+        columns=columns,
+        dtypes=dtypes,
+        data=df.values.tolist(),
+    )
 
 
 class DataFrame(PydanticBaseModel):
-    index: list | None = Field(None)
+    index: list[int] | list[str] | None = Field(None)
     columns: list[str] | None
     dtypes: list[str] | None
-    data: list | None
+    # TODO The order is important here at the moment, in particular having int before
+    # float! This should likely not be the case, but using StrictInt and StrictFloat
+    # from pydantic only created even more errors.
+    data: (
+        list[
+            list[
+                bool
+                | datetime
+                | int
+                | float
+                | str
+                | dict[str, Any]
+                # TODO should be able to remove this once PR#122 is merged
+                | list[float | int | str]
+                | None
+            ]
+        ]
+        | None
+    )
```

Comment:
This type hint is not great as explained in the comment. I think I really don't like pydantic using the order in type hints to coerce types, because that's something that I wouldn't expect. Here, it seems to be important that `int` come before `float`. Similar to `meta`, I tried using `StrictInt` etc, but this broke some other tests. Please let me know if you have a good idea for this.

Reply:
I would also guess that this worsens performance for the api client significanlty, did you run the benchmarks before and after? It will now check /every single value/ and try to coerce it, no?

Reply:
There do seem to be differences, but I find it hard to put my finger on them. The `filter_` tests seem to run a little faster on `main`, but the other tests seem to be a lot faster on this branch.

Reply:
Hmm, so I tested my theory specifically and it seems to be true, not sure why it is not showing up in benchmarks here, maybe its a hint that its actually not such a big deal...

I took a few minutes to write this benchmark:

```py
import pytest
from pydantic import BaseModel

class Flat(BaseModel):
    data: list


class Detailed(BaseModel):
    data: list[int]


@pytest.mark.parametrize(
    "data",
    [
        list(range(1_000_000)),
        list(range(10_000_000)),
        list(range(100_000_000)),
    ],
)
def test_flat(benchmark, data):
    def bench():
        return Flat(data=data)
    
    benchmark(bench)



@pytest.mark.parametrize(
    "data",
    [
        list(range(1_000_000)),
        list(range(10_000_000)),
        list(range(100_000_000)),
    ],
)
def test_detailed(benchmark, data):
    def bench():
        return Detailed(data=data)
    
    benchmark(bench)
```
... which benchmarks a `list` type-hint versus `list[int]`.
The results are a little bit surprising:

```
------------------------------------------------------------------------------------------ benchmark: 6 tests -----------------------------------------------------------------------------------------
Name (time in ms)               Min                   Max                  Mean             StdDev                Median                IQR            Outliers       OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_flat[data0]             4.7867 (1.0)          6.0222 (1.0)          4.9039 (1.0)       0.1517 (1.17)         4.8734 (1.0)       0.0843 (1.0)          8;11  203.9214 (1.0)         149           1
test_detailed[data0]        17.1237 (3.58)        17.6601 (2.93)        17.3089 (3.53)      0.1298 (1.0)         17.2635 (3.54)      0.1706 (2.02)         19;2   57.7737 (0.28)         61           1
test_flat[data1]            65.7662 (13.74)       67.2023 (11.16)       66.1108 (13.48)     0.4609 (3.55)        65.9060 (13.52)     0.2877 (3.41)          3;3   15.1261 (0.07)         15           1
test_detailed[data1]       176.7581 (36.93)      180.5254 (29.98)      178.1051 (36.32)     1.4764 (11.38)      177.5801 (36.44)     2.2640 (26.85)         1;0    5.6147 (0.03)          6           1
test_flat[data2]           666.2161 (139.18)     674.3390 (111.98)     669.3321 (136.49)    3.3201 (25.58)      669.2435 (137.33)    4.9705 (58.95)         1;0    1.4940 (0.01)          5           1
test_detailed[data2]     1,797.9444 (375.62)   1,870.2783 (310.57)   1,817.4652 (370.62)   29.8766 (230.20)   1,808.0563 (371.00)   23.5364 (279.15)        1;1    0.5502 (0.00)          5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
**Both versions** scale with the length of the list, which I did not expect since in my head the "Flat" class should validate in constant time. (i now also wish i didnt call it "Flat")
Thats probably why it doesnt make a difference in light of the apparent performance improvements in this PR. (Do we know what these are?)